File: 13/04/Tests/BaseTest.cs
using System;
using System.Configuration;
using OpenQA.Selenium;
using NUnit.Framework;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Remote;
using NUnit.Framework.Interfaces;

namespace Tests
{
    [TestFixture]
    class BaseTest
    {
        protected IWebDriver Driver;
        public static string ApplicationBaseUrl;
        private static string VendorDirectory;
        private static string BrowserName;
        private static string Host;
        private static string BrowserVersion;
        private static string Platform;

        private void LoadConfigValues()
        {
            var configReader    = new AppSettingsReader();
            Host                = (string)configReader.GetValue("Host", typeof(string));
            BrowserName         = (string)configReader.GetValue("BrowserName", typeof(string));
            BrowserVersion      = (string)configReader.GetValue("BrowserVersion", typeof(string));
            Platform            = (string)configReader.GetValue("Platform", typeof(string));
            ApplicationBaseUrl  = (string)configReader.GetValue("ApplicationBaseUrl", typeof(string));
            VendorDirectory     = System.IO.Directory.GetParent(
                                    System.IO.Path.GetDirectoryName(
                                    typeof(Tests.BaseTest).Assembly.Location)).
                                    Parent.FullName + @"\Vendor";
        }

        [SetUp]
        protected void SetUp()
        {
            LoadConfigValues();
            switch (Host.ToLower())
            {
                case "localhost":
                    switch (BrowserName.ToLower())
                    {
                        case "firefox":
			    var VendorDirectory = System.IO.Directory.GetParent(
						    System.AppDomain.CurrentDomain.BaseDirectory).
							Parent.Parent.FullName
							    + @"\Vendor";
			    var Service = FirefoxDriverService.CreateDefaultService(VendorDirectory);
			    Driver = new FirefoxDriver(Service);
                            break;
                        case "chrome":
                            Driver = new ChromeDriver(VendorDirectory);
                            break;
                    }
                    break;
                case "saucelabs":
                    DesiredCapabilities caps = new DesiredCapabilities();
                    caps.SetCapability(CapabilityType.BrowserName, BrowserName);
                    caps.SetCapability(CapabilityType.Version, BrowserVersion);
                    caps.SetCapability(CapabilityType.Platform, Platform);
                    caps.SetCapability("username", System.Environment.GetEnvironmentVariable("SAUCE_USERNAME"));
                    caps.SetCapability("accessKey", System.Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY"));
                    caps.SetCapability("name", TestContext.CurrentContext.Test.Name);
                    Driver = new RemoteWebDriver(new Uri("http://ondemand.saucelabs.com:80/wd/hub"), caps);
                    break;
            }
        }

        [TearDown]
        protected void TearDown()
        {
            if (Host.Equals("saucelabs"))
                {
                    bool testPassed = TestContext.CurrentContext.Result.Outcome.Status.Equals(TestStatus.Passed);
                try
                {
                    ((IJavaScriptExecutor)Driver).ExecuteScript("sauce:job-result=" + (testPassed ? "passed" : "failed"));
                    Console.WriteLine("https://saucelabs.com/beta/tests/" + ((RemoteWebDriver)Driver).SessionId);
                }    
                finally
                {
                    Driver.Quit();
                }
            } else
            {
                Driver.Quit();
            }
        }
    }
}

File: java/Quickstart/Mod3/3.05/src/test/java/tests/BaseTest.java
// filename: tests/BaseTest.java
package tests;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.junit.Rule;
import org.junit.rules.ExternalResource;
import org.openqa.selenium.MutableCapabilities;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.RemoteWebDriver;


import java.net.URL;

import static tests.Config.*;


public class
BaseTest {

    protected WebDriver driver;

    @Rule
    public ExternalResource resource = new ExternalResource() {
        @Override
        protected void before() throws Exception {
            switch (host) {
                case "saucelabs": {
                    String sauceUrl = "https://ondemand.us-west-1.saucelabs.com/wd/hub";
                    MutableCapabilities sauceOptions = new MutableCapabilities();
                    sauceOptions.setCapability("username", sauceUser);
                    sauceOptions.setCapability("accesskey", sauceKey);
                    MutableCapabilities capabilities = new MutableCapabilities();
                    capabilities.setCapability("browserName", browserName);
                    capabilities.setCapability("browserVersion", browserVersion);
                    capabilities.setCapability("platformName", platformName);
                    capabilities.setCapability("sauce:options", sauceOptions);
                    driver = new RemoteWebDriver(new URL(sauceUrl), capabilities);
                    break;
                }
                case "localhost": {
                    if ("firefox".equals(browserName)) {
                        WebDriverManager.firefoxdriver().setup();
                        driver = new FirefoxDriver();
                    } else if ("chrome".equals(browserName)) {
                        WebDriverManager.chromedriver().setup();
                        driver = new ChromeDriver();
                    }
                    break;
                }
            }
        }
        @Override
        protected void after() {
            driver.quit();
        }
        };
    };

